### ISSUE TITLE
Poor CPU scaling; NDArray.asnumpy bottleneck?

### ISSUE THREAD

**Issue by novog** [apache/incubator-mxnet#4552]
I am having a performance scaling issue in CPU context. From what I can tell, it looks like this issue could suggest a bottleneck in the call NDArray.asnumpy(). I initially noticed this problem using the old FeedForward class, and more recently duplicated it using the newer Module class; all details below are using Module.

When switching MXNET_CPU_WORKER_NTHREADS and OMP_NUM_THREADS from 1 to 4 on a 4-core machine, I only get a 13% training speed improvement. My CPU usage does increase about threefold, so the settings would seem to be working. Also, I get a 268% speed improvement in a numpy .dot operation, so I do see improvements there.

Given that multithreading seems to be working but there isn't much performance improvement, I tried using cProfile to identify any potential bottlenecks. Using a test function (code below), here are the most expensive calls:

    ncalls  tottime  percall  cumtime  percall filename:lineno(function)
         1    0.011    0.011   11.299   11.299 <string>:1(<module>)
         1    0.000    0.000   10.997   10.997 base_module.py:287(fit)
        21    0.000    0.000   10.752    0.512 module.py:521(update_metric)
        21    0.000    0.000   10.752    0.512 executor_group.py:413(update_metric)
        21    0.002    0.000   10.751    0.512 metric.py:134(update)
        42   10.745    0.256   10.748    0.256 ndarray.py:488(asnumpy)

So, it looks like the program is spending almost all of its time in the two calls to asnumpy() made in the metric's update function. Can anyone explain this effect?

My environment:
Operating System: Windows 7 SP1
Compiler: Visual Studio 2015
Package used (Python/R/Scala/Julia): Python
Multi-threading library used: OpenBLAS
MXNet commit hash (`git rev-parse HEAD`): 440b6379708693c5e8c1537b0ded65be9f2011bb (from December 30th)
Python version and distribution: WinPython 2.7.10 64-bit

Code for numpy array multiplication test:

    def numpy_test():
        A = np.random.rand(4000, 4000)
        B = np.random.rand(4000, 4000)
        C = A.dot(B)

Sample code for mxnet test:

    def asnumpy_test(hidden_layer_count):
        X = np.random.rand(10000, 1000)
        y = np.random.randint(2, size=(10000,))
        train_iter = mx.io.NDArrayIter(X, y, 500)
        X_eval = np.random.rand(500, 1000)
        y_eval = np.random.randint(2, size=(500,))
        eval_iter = mx.io.NDArrayIter(X_eval, y_eval, 500)

        net = mx.sym.Variable('data')
        for i in range(hidden_layer_count):
            layer = mx.sym.FullyConnected(data=net, num_hidden=20)
            layer = mx.sym.LeakyReLU(data=layer, act_type='prelu')
            layer = mx.sym.BatchNorm(data=layer)
            net = mx.sym.Concat(net, layer, num_args=2)
        label = mx.sym.Variable('softmax_label')
        net = mx.sym.SoftmaxOutput(data=net, label=label)
    
        model = mx.mod.Module(net, ['data'], ['softmax_label'])
        model.bind(train_iter.provide_data, train_iter.provide_label)
        model.init_params(mx.init.MSRAPrelu())
        optimizer_params = {'learning_rate':0.1, 'momentum':0.9, 'wd':0.0001}
        model.fit(train_iter, eval_iter, 'acc', optimizer_params=optimizer_params, num_epoch=1)

**Comment by piiswrong** [comment]
try USE_MKL2017 = 1 in config.mk

**Comment by novog** [comment]
OK, I did a build using the Intel libraries for OpenCV, numpy, and MXNet, setting USE_MKL2017 to 1. As I increase the environment variable MKL_NUM_THREADS, I see CPU usage increase proportionately in both test, until it peaks at the number of physical cores on the machine. (Additionally changing MXNET_CPU_WORKER_NTHREADS does not seem to have an additional effect, either on speed or CPU usage.) So, the Intel multithreading libraries appear to be working.

Once again, I see a substantial (183%) increase in speed for the numpy test, but only a 17% improvement in speed in the mxnet test. The cProfile results are similar (although MKL was a bit faster generally than OpenBLAS for the MXNet test on this machine). The vast majority of the time still appears to be spent in the NDArray.asnumpy() call:

       ncalls  tottime  percall  cumtime  percall filename:lineno(function)
            1    0.014    0.014    9.763    9.763 <string>:1(<module>)
            1    0.000    0.000    9.460    9.460 base_module.py:287(fit)
           21    0.000    0.000    9.155    0.436 module.py:521(update_metric)
           21    0.000    0.000    9.155    0.436 executor_group.py:413(update_metric)
           21    0.002    0.000    9.153    0.436 metric.py:134(update)
           42    9.147    0.218    9.149    0.218 ndarray.py:488(asnumpy)

Is there any other information that I can provide, or any other suggestions? Thanks!

**Comment by novog** [comment]
Closed accidentally.

**Comment by yajiedesign** [comment]
This issue is closed due to lack of activity in the last 90 days. Feel free to reopen if this is still an active issue. Thanks!
